lib/Dialect/SV/Transforms/HWDeadCodeElim.cpp

```diff
@@ -0,0 +1,397 @@
+//===- HWDeadCodeElim.cpp - Elimination of dead ports ---------------------===//
+//
+// Proprietary and Confidential Software of SiFive Inc. All Rights Reserved.
+// See the LICENSE file for license information.
+// SPDX-License-Identifier: UNLICENSED
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the `HWDeadCodeElim` pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "circt/Dialect/HW/HWInstanceGraph.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "hw-dead-code-elim"
+
+using namespace mlir;
+using namespace circt;
+
+using llvm::SmallBitVector;
+
+//===----------------------------------------------------------------------===//
+// Liveness analysis for dead-code elimination
+//===----------------------------------------------------------------------===//
+
+namespace {
+class Liveness {
+public:
+  Liveness(ModuleOp circuit, hw::InstanceGraph &instanceGraph)
+      : circuit(circuit), instanceGraph(instanceGraph) {}
+
+  LogicalResult run();
+
+  // Return true if a value has users and should be preserved.
+  bool isLive(Value value) { return liveValues.count(value); }
+
+private:
+  void markLive(Value value);
+  void markLive(Operation *op);
+
+  void propagate(Value result, sv::WireOp wire);
+  void propagate(Value result, hw::InstanceOp inst);
+
+private:
+  ModuleOp circuit;
+  hw::InstanceGraph &instanceGraph;
+
+  DenseSet<Value> liveValues;
+  SmallVector<Value> q;
+};
+} // namespace
+
+// Returns true if an op can be deleted.
+static bool canBeDeleted(Operation *op) {
+  if (isa<sv::AssignOp>(op))
+    return true;
+  if (op->hasAttr("inner_sym"))
+    return false;
+  if (op->hasTrait<sv::ProceduralOp>())
+    return false;
+  if (op->hasTrait<sv::NonProceduralOp>())
+    return false;
+  return mlir::MemoryEffectOpInterface::hasNoEffect(op);
+}
+
+LogicalResult Liveness::run() {
+  for (auto module : circuit.getOps<hw::HWModuleOp>()) {
+    // Mark all public module input and output ports as live.
```

Comment:
normally this kind of thing would be demand-based, and an input is demanded.  It might not be deletable, but it shouldn't be live.